applications/serial_lte_modem/src/slm_ppp.c

```diff
@@ -0,0 +1,309 @@
+static K_THREAD_STACK_DEFINE(ppp_data_passing_thread_stack, KB(2));
+static void ppp_data_passing_thread(void*, void*, void*);
+
+/* We use only the default PDP context. */
+enum { PDP_CID = 0 };
+
+enum {
+	ZEPHYR_FD_IDX, /* Raw Zephyr socket to pass data to/from the PPP link. */
+	MODEM_FD_IDX, /* Raw modem socket to pass data to/from the LTE link. */
+	PPP_FDS_COUNT
+};
+const char *const ppp_socket_names[PPP_FDS_COUNT] = {
+	[ZEPHYR_FD_IDX] = "Zephyr",
+	[MODEM_FD_IDX] = "modem"
+};
+static int ppp_fds[PPP_FDS_COUNT] = { -1, -1 };
+
+static bool open_ppp_sockets(void)
+{
+	int ret;
+
+	ppp_fds[ZEPHYR_FD_IDX] = socket(AF_PACKET, SOCK_RAW | SOCK_NATIVE, IPPROTO_RAW);
+	if (ppp_fds[ZEPHYR_FD_IDX] < 0) {
+		LOG_ERR("Zephyr socket creation failed (%d).", errno);
+		return false;
+	}
+
+	ppp_zephyr_dst_addr = (struct sockaddr_ll){
+		.sll_family = AF_PACKET,
+		.sll_ifindex = ppp_iface_idx
+	};
+	ret = bind(ppp_fds[ZEPHYR_FD_IDX],
+		   (const struct sockaddr *)&ppp_zephyr_dst_addr, sizeof(ppp_zephyr_dst_addr));
+	if (ret < 0) {
+		LOG_ERR("Failed to bind Zephyr socket (%d).", errno);
+		return false;
+	}
+
+	ppp_fds[MODEM_FD_IDX] = socket(AF_PACKET, SOCK_RAW, 0);
+	if (ppp_fds[MODEM_FD_IDX] < 0) {
+		LOG_ERR("Modem socket creation failed (%d).", errno);
+		return false;
+	}
+
+	return true;
+}
+
+static void close_ppp_sockets(void)
+{
+	for (size_t i = 0; i != ARRAY_SIZE(ppp_fds); ++i) {
+		if (ppp_fds[i] < 0) {
+			continue;
+		}
+		if (close(ppp_fds[i])) {
+			LOG_WRN("Failed to close %s socket (%d).",
+				ppp_socket_names[i], errno);
+		}
+		ppp_fds[i] = -1;
+	}
+}
+
+static bool configure_ppp_link_ip_addresses(struct ppp_context *ctx)
+{
+	char addr4[INET_ADDRSTRLEN];
+	char addr6[INET6_ADDRSTRLEN];
+
+	util_get_ip_addr(PDP_CID, addr4, addr6);
+
+	if (*addr4) {
+		if (inet_pton(AF_INET, addr4, &ctx->ipcp.my_options.address) != 1) {
+			return false;
+		}
+	} else if (!*addr6) {
+		LOG_ERR("No connectivity.");
+		return false;
+	}
+	if (*addr6) {
+		struct in6_addr in6;
+		struct ipv6cp_options *const opts = &ctx->ipv6cp.my_options;
+
+		if (inet_pton(AF_INET6, addr6, &in6) != 1) {
+			return false;
+		}
+		/* The interface identifier is the last 64 bits of the IPv6 address. */
+		BUILD_ASSERT(sizeof(in6) >= sizeof(opts->iid));
+		/* Setting it doesn't seem to change anything in practice, but this
+		 * is what it's meant for, so fill it anyway.
+		 */
+		memcpy(opts->iid, (uint8_t *)(&in6 + 1) - sizeof(opts->iid), sizeof(opts->iid));
+	}
+	return true;
+}
+
+static int slm_ppp_start(void)
+{
+	int ret;
+	unsigned int mtu;
+	struct ppp_context *const ctx = net_if_l2_data(ppp_iface);
+
+	if (!configure_ppp_link_ip_addresses(ctx)) {
+		return -EADDRNOTAVAIL;
+	}
+
+	ret = pdn_dynamic_params_get(PDP_CID, &ctx->ipcp.my_options.dns1_address,
+				     &ctx->ipcp.my_options.dns2_address, &mtu);
+	if (ret) {
+		return ret;
+	}
+
+	/* Set the PPP MTU to that of the LTE link. */
+	mtu = MIN(mtu, sizeof(ppp_data_buf));
```

Comment:
Just a note that in some networks (either Elisa or DNA sim that I have), the MTU returned by CGCONTRDP was problematic when trying to send full size UDP packages. The actual transmittable package was much less than what this returned (headers were taken into account).

Reply:
Interesting. Do you recall the numbers? We might have to make the MTU configurable at some point, or something of the sort.

Reply:
Unfortunately not and anyway it differed between operators.